tarsier/core.py

```diff
@@ -27,24 +29,48 @@ def __init__(self, ocr_service: OCRService):
 
     async def page_to_image(
         self, driver: AnyDriver, tag_text_elements: bool = False, tagless: bool = False
-    ) -> Tuple[bytes, Dict[int, str]]:
+    ) -> Tuple[bytes, bytes, Dict[int, str]]:
         adapter = adapter_factory(driver)
+        initial_screenshot = await self._take_screenshot(adapter)
         tag_to_xpath = (
             await self._tag_page(adapter, tag_text_elements) if not tagless else {}
         )
-        screenshot = await self._take_screenshot(adapter)
+        await self._hide_non_tag_elem(adapter)
+        tagged_screenshot = await self._take_screenshot(adapter)
+        await self._revert_visibilities(adapter)
         if not tagless:
             await self._remove_tags(adapter)
-        return screenshot, tag_to_xpath if not tagless else {}
+        return (
+            initial_screenshot,
+            tagged_screenshot,
+            tag_to_xpath if not tagless else {},
+        )
 
     async def page_to_text(
         self, driver: AnyDriver, tag_text_elements: bool = False, tagless: bool = False
     ) -> Tuple[str, TagToXPath]:
-        image, tag_to_xpath = await self.page_to_image(
+        untagged_image, tagged_image, tag_to_xpath = await self.page_to_image(
             driver, tag_text_elements, tagless
         )
-        page_text = self._run_ocr(image)
-        return page_text, tag_to_xpath
+        untagged_ocr_annotations = self._ocr_service.annotate(untagged_image)
+        tagged_ocr_annotations = self._ocr_service.annotate(tagged_image)
+
+        combined_annotations: ImageAnnotatorResponse = {
+            "words": untagged_ocr_annotations["words"] + tagged_ocr_annotations["words"]
+        }
+        combined_annotations["words"] = list(
+            sorted(
+                combined_annotations["words"],
+                key=lambda x: (
+                    x["midpoint_normalized"][1],
+                    x["midpoint_normalized"][0],
+                ),
+            )
+        )
```

Comment:
Should maybe make a combine_annotations() method that handles this resorting (To decouple this function having to care that its sorted)